crates/bevy_asset/macros/src/lib.rs

```diff
@@ -41,33 +41,101 @@ fn derive_dependency_visitor_internal(
+                            #bevy_asset_path::VisitAssetDependencies::visit_dependencies(&self.#field_ident, visit);
+                        });
+                    }
                 }
             }
+
+            quote! {
+                #(#field_visitors)*
+            }
         }
-    } else {
-        return Err(syn::Error::new(
-            Span::call_site().into(),
-            "Asset derive currently only works on structs",
-        ));
-    }
+        Data::Enum(data_enum) => {
+            let mut any_case_required = false;
+            let cases = data_enum.variants.iter().map(|variant| {
+                let mut any_member_omitted = false;
+                let mut field_names = Vec::new();
+                let mut field_visitors = Vec::new();
 
-    let struct_name = &ast.ident;
-    let (impl_generics, type_generics, where_clause) = &ast.generics.split_for_impl();
+                for (i, field) in variant.fields.iter().enumerate() {
+                    if field
+                        .attrs
+                        .iter()
+                        .any(|a| a.path().is_ident(DEPENDENCY_ATTRIBUTE))
+                    {
+                        if let Some(field_ident) = &field.ident {
+                            field_names.push(quote!{ #field_ident });
+                            field_visitors.push(quote! {
+                                #bevy_asset_path::VisitAssetDependencies::visit_dependencies(#field_ident, visit);
+                            });
+                        } else {
+                            let name = format_ident!("member_{i}");
+                            field_names.push(quote!{ #name });
+                            field_visitors.push(quote! {
+                                #bevy_asset_path::VisitAssetDependencies::visit_dependencies(#name, visit);
+                            });
+                        }
+                    } else {
+                        any_member_omitted = true;
+                    }
+                }
+
+                if any_member_omitted {
+                    field_names.push(quote!{ .. });
+                }
+
+                let field_data = match variant.fields {
+                    syn::Fields::Named(_) => quote! { {#(#field_names,)*} },
+                    syn::Fields::Unnamed(_) => quote! { (#(#field_names,)*) },
+                    syn::Fields::Unit => quote! { },
+                };
+
+                any_case_required |= !field_visitors.is_empty();
+
+                let var_ident = &variant.ident;
+                quote! {
+                    Self::#var_ident #field_data => {
+                        #(#field_visitors)*
+                    }
+                }
+            }).collect::<Vec<_>>();
+
+            if !any_case_required {
+                quote! {}
+            } else {
+                quote! {
+                    match self {
+                        #(#cases)*
+                    }
+                }
+            }
```

Comment:
```suggestion
            any_case_required.then(|| quote!(match self { #(#cases)* }))
```

Reply:
```
119 | |              any_case_required.then(|| quote!(match self { #(#cases)* }))
    | |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `TokenStream`, found `Option<TokenStream>`
```

Reply:
+ unwrap_or_default works, but honestly i find it harder to read

Reply:
It should work, try wrapping the other `Data::Struct` branch return value in a `Some(_)`